xpdan/data_reduction.py

```diff
@@ -0,0 +1,360 @@
+                if label is not None:
+                    feature_list.append(str(label))
+                else:
+                    pass
+        # get fields
+        for key in self.fields:
+            el = str(run_start.get(key, None))
+            if el is not None:
+                # truncate string length
+                if len(el) >12:
+                    value = el[:12]
+                # clear space
+                feature = _clean_info(el)
+                feature_list.append(feature)
+            else:
+                pass
+        # get data fields
+        for key in self.data_fields:
+            val = event['data'].get(key, None)
+            if el is not None:
+                feature = "{}={}".format(key, val)
+                feature_list.append(feature)
+            else:
+                pass
+        # get uid
+        feature_list.append(uid)
+        return "_".join(feature_list)
+
+    def pull_dark(self, header):
+        dark_uid = header.start.get(an_glbl.dark_field_key, None)
+        if dark_uid is None:
+            print("INFO: no dark frame is associated in this header, "
+                  "subrraction will not be processed")
+            return None
+        else:
+            dark_search = {'group': 'XPD', 'uid': dark_uid}
+            dark_header = db(**dark_search)
+            dark_img = np.asarray(get_images(dark_header,
+                                             self.image_field)).squeeze()
+        return dark_img, dark_header[0].start.time
+
+    def _dark_sub(self, event, dark_img):
+        """ priviate method operates on event level """
+        dark_sub = False
+        img = event['data'][self.image_field]
+        if dark_img is not None and isinstance(dark_img, np.ndarray):
+            dark_sub = True
+            img -= dark_img
+        ind = event['seq_num']
+        event_timestamp = event['timestamps'][self.image_field]
+        return img, event_timestamp, ind, dark_sub
+
+    def dark_sub(self, header):
+        """ public method operates on header level """
+        img_list = []
+        timestamp_list = []
+        dark_img, dark_time_stamp = self.pull_dark(header)
+        for ev in get_events(header, fill=True):
+            sub_img, timestamp, ind, dark_sub = self._dark_sub(ev, dark_img)
+            img_list.append(sub_img)
+            timestamp_list.append(timestamp)
+        return img_list, timestamp_list, dark_img, header.start
+
+    def _file_name(self, event, event_timestamp, ind):
+        """ priviate method operates on event level """
+        f_name = self._feature_gen(event)
+        f_name = '_'.join([f_name,
+                           _timestampstr(event_timestamp)])
+        f_name = '{}_{:04d}.tif'.format(f_name, ind)
+        return f_name
+
+
+# init
+xpd_data_proc = DataReduction()
+ai = AzimuthalIntegrator()
+
+
+### analysis function operates at header level ###
+def _prepare_header_list(headers):
+    if type(list(headers)[1]) == str:
```

Comment:
I guess I am a bit confused what the header is?
I thought the header was the run start header, which returned False when asked if it is a list.